designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+  * This appoach does not enforce naming of plugins but enforces the directories the plugins have to be in, which is a more common and typical approach.
+
+  * The one-plugin-per-directory requirement eases creation of a plugin bundle for sharing.
+
+* Cons
+  * This makes it mandatory for every plugin that is specified in the `layout` field to have additional configuration information such as Group, Version and Kind in the `plugins` field in the PROJECT file for it to be discoverable. Currently, the `plugins` field is more like an optional configuration storage field.
+
+#### Approach #3
+
+Another approach I would like to mention is adding plugin executables with a prefix `kubebuilder-` followed by the plugin name to the PATH variable. This will enable KB to traverse through the PATH looking for the plugin executables starting with the prefix `kubebuilder-` and matching by the plugin name that was provided in the CLI. Furthermore, a check should be added to verify that the match is an executable or not and return an error if it's not an executable. This approach provides a lot of flexibility in terms of plugin discovery as all the user needs to do is to add the plugin executable to the PATH and Kubebuilder will discover it.
+
+* Pros
+  * `kubectl` and `git` follow the same approach for discovering plugins, so there’s prior art.
+
+  * There’s a lot of flexibility in just dropping plugin binaries to PATH variable and enabling the discovery without having to enforce any other constraints on the placements of the plugins.
+
+* Cons
+  * Enumerating the list of all available plugins might be a bit tough compared to having a single folder with the list of available plugins and having to enumerate those.
+
+  * These plugin binaries cannot be run in a standalone manner outside of Kubebuilder, so may not be very ideal to add them to the PATH var.
+
+#### What Plugin system should we use
+
+Based on the evaluation of plugin libraries such as the [built-in go-plugin library](https://golang.org/pkg/plugin/) and [Hashicorp’s plugin library](https://github.com/hashicorp/go-plugin), it is more suitable to write our own custom plugin library as the built-in plugin library seems to be more suitable for in-tree plugins rather than out-of-tree plugins and it doesn’t offer cross language support, thereby making it a non-starter. Hashicorp’s go plugin system is more suitable than the built-in go-plugin library as it enables cross language/platform support. However, it is more suited for long running plugins as opposed to short lived plugins and the usage of protobuf could be overkill as we will not be handling 10s of 1000s of deserializations.
+
+For the stated reasons, I propose we use our own plugin system that passes JSON blobs back and forth across stdin/stdout and make stdin/stdout the only default option for now as it’s the most language agnostic way and it makes it easy to work with most languages.
```

Comment:
I would flip these two paragraphs. I would start with:

I propose we use our own plugin system ...

We came to this conclusion after evaluating plugin libraries such as the built-ing go-plugin library and Hashicorp's plugin library ...